### DISCUSSION TITLE
Spoofing Player Input RPC's

### DISCUSSION THREAD

**Discussion by D1GQ** [winstxnhdw/lc-hax#477] [Ideas]
# Overview
I was thinking since the player connects to the player object when joining a game, is it possible to do something known as Puppet Master Hacks, the idea is spoofing server rpcs to tell the server to update the players location or rotation for other players, so pretty much either taking control of the player temporarily or making it look like they're hacking.
# Code
This is one part of the PlayerControllerB that I am looking into:
```
[ServerRpc]
private void UpdatePlayerRotationServerRpc(short newRot, short newYRot)
{
	NetworkManager networkManager = base.NetworkManager;
	if (networkManager == null || !networkManager.IsListening)
	{
		return;
	}
	if (this.__rpc_exec_stage != NetworkBehaviour.__RpcExecStage.Server && (networkManager.IsClient || networkManager.IsHost))
	{
		if (base.OwnerClientId != networkManager.LocalClientId)
		{
			if (networkManager.LogLevel <= Unity.Netcode.LogLevel.Normal)
			{
				Debug.LogError("Only the owner can invoke a ServerRpc that requires ownership!");
			}
			return;
		}
		ServerRpcParams serverRpcParams;
		FastBufferWriter writer = base.__beginSendServerRpc(588787670U, serverRpcParams, RpcDelivery.Reliable);
		BytePacker.WriteValueBitPacked(writer, newRot);
		BytePacker.WriteValueBitPacked(writer, newYRot);
		base.__endSendServerRpc(ref writer, 588787670U, serverRpcParams, RpcDelivery.Reliable);
	}
	if (this.__rpc_exec_stage != NetworkBehaviour.__RpcExecStage.Server || (!networkManager.IsServer && !networkManager.IsHost))
	{
		return;
	}
	try
	{
		this.UpdatePlayerRotationClientRpc(newRot, newYRot);
	}
	catch (Exception arg)
	{
		Debug.Log(string.Format("Client rpc parameters were likely not correct, so an RPC was skipped: {0}", arg));
	}
}
```

**Comment by winstxnhdw** [comment]
I feel like this is verified by the server. A long time back when I didn’t know how the RPCs work yet, I was moving everyone everywhere but it seems only I can see it.

**Reply by D1GQ** [reply]
That's most likely true, but there is a chance that you were updating the positions locally on your client, A lot of server RPC's at the end call a client RPC.

**Reply by winstxnhdw** [reply]
Possible. Maybe someone can test this.
